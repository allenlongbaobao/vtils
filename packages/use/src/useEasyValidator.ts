import {EasyValidator, EasyValidatorData, EasyValidatorRules, EasyValidatorValidateReturn} from 'vtils'
import {useRef} from 'react'

export type UseEasyValidatorValidateResult<D extends EasyValidatorData> = EasyValidatorValidateReturn<D> & { data: D }

export interface UseEasyValidatorValidate<D extends EasyValidatorData> {
  (): Promise<UseEasyValidatorValidateResult<D>>,
}

export interface UseEasyValidatorReturn<D extends EasyValidatorData> {
  /**
   * 验证数据。
   */
  validate: UseEasyValidatorValidate<D>,
}

/**
 * 数据校验器。
 *
 * @param data 数据
 * @param rules 校验规则
 * @returns 返回包含验证函数的对象
 * @example
 * ```ts
 * const [name, setName] = useState('')
 * const [pass, setPass] = useState('')
 * const ev = useEasyValidator({ name, pass }, [
 *   {
 *     key: 'name',
 *     required: true,
 *     message: '姓名不能为空',
 *   },
 *   {
  *     key: 'pass',
  *     test: data => data.pass.length >= 6,
  *     message: '密码至少应为6位',
  *   },
 * ])
 * const handleRegisterClick = useCallback(() => {
 *   ev.validate().then(res => {
 *     if (res.valid) {
 *       console.log(res.data)
 *     } else {
 *       console.log(res.unvalidRules[0].message)
 *     }
 *   })
 * }, [])
 * ```
 */
export function useEasyValidator<D extends EasyValidatorData>(data: D, rules: EasyValidatorRules<D>): UseEasyValidatorReturn<D> {
  const store = useRef<{
    data: D,
    easyValidator: EasyValidator<D>,
    validate: UseEasyValidatorValidate<D>,
  }>({} as any)

  store.current.data = data

  if (!store.current.easyValidator) {
    store.current.easyValidator = new EasyValidator<D>(rules)
    store.current.validate = () => {
      const currentData = store.current.data
      return store.current.easyValidator.validate(currentData).then(res => {
        (res as any).data = currentData
        return res as any
      })
    }
  }

  return store.current
}
