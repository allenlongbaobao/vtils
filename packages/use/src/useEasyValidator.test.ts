import {act, renderHook} from '@testing-library/react-hooks'
import {useCallback, useState} from 'react'
import {useEasyValidator} from './useEasyValidator'
import {wait} from 'vtils'

test('表现正常', async () => {
  let evResult: any

  const {result} = renderHook(() => {
    const [name, setName] = useState('')
    const [pass, setPass] = useState('')
    const ev = useEasyValidator({name, pass}, [
      {
        key: 'name',
        required: true,
        message: '姓名不能为空',
      },
      {
        key: 'pass',
        test: data => data.pass.length >= 6,
        message: '密码至少应为6位',
      },
    ])
    const handleRegisterClick = useCallback(() => {
      ev.validate().then(res => {
        evResult = res
      })
    }, [])
    return {setName, setPass, handleRegisterClick}
  })

  result.current.handleRegisterClick()
  await wait(10)
  expect(evResult).toMatchSnapshot('姓名、密码皆为空')

  act(() => result.current.setName('方方'))
  result.current.handleRegisterClick()
  await wait(10)
  expect(evResult).toMatchSnapshot('姓名不为空、密码为空')

  act(() => result.current.setPass('123'))
  result.current.handleRegisterClick()
  await wait(10)
  expect(evResult).toMatchSnapshot('姓名、密码都不为空，但密码少于6位')

  act(() => result.current.setPass('123456'))
  result.current.handleRegisterClick()
  await wait(10)
  expect(evResult).toMatchSnapshot('姓名、密码都不为空，且密码等于6位')
})
