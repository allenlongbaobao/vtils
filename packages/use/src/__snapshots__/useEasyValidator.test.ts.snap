// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`表现正常: 姓名、密码皆为空 1`] = `
Object {
  "data": Object {
    "name": "",
    "pass": "",
  },
  "messages": Object {
    "name": "姓名不能为空",
    "pass": "密码至少应为6位",
  },
  "unvalidRules": Array [
    Object {
      "key": "name",
      "message": "姓名不能为空",
      "required": true,
    },
    Object {
      "key": "pass",
      "message": "密码至少应为6位",
      "test": [Function],
    },
  ],
  "valid": false,
}
`;

exports[`表现正常: 姓名、密码都不为空，且密码等于6位 1`] = `
Object {
  "data": Object {
    "name": "方方",
    "pass": "123456",
  },
  "messages": Object {},
  "unvalidRules": Array [],
  "valid": true,
}
`;

exports[`表现正常: 姓名、密码都不为空，但密码少于6位 1`] = `
Object {
  "data": Object {
    "name": "方方",
    "pass": "123",
  },
  "messages": Object {
    "pass": "密码至少应为6位",
  },
  "unvalidRules": Array [
    Object {
      "key": "pass",
      "message": "密码至少应为6位",
      "test": [Function],
    },
  ],
  "valid": false,
}
`;

exports[`表现正常: 姓名不为空、密码为空 1`] = `
Object {
  "data": Object {
    "name": "方方",
    "pass": "",
  },
  "messages": Object {
    "pass": "密码至少应为6位",
  },
  "unvalidRules": Array [
    Object {
      "key": "pass",
      "message": "密码至少应为6位",
      "test": [Function],
    },
  ],
  "valid": false,
}
`;
